py-polars/src/lazyframe/visit.rs

```diff
@@ -133,8 +133,15 @@ impl NodeTraverser {
             // CSE expressions for to_field to work with expressions
             // that reference them, but is not part of the public
             // schema of the node.
-            let schema = ir_node.schema(&lp_arena);
+            let schema = match ir_node {
+                // Select operates on the input schema
+                IR::Select { .. } => ir_node.input_schema(&lp_arena).unwrap(),
```

Comment:
Everything operates on it's input schema? Shouldn't we take `input_schema` for all?

Reply:
Ah yes, I think you are right. I confused myself thinking through various scenarios.

Specifically, join had me worried because the left (right) keys have dtypes that depend on the schema context of the left (right) input, but the correct way to treat that is to ensure that the relevant input is "active" when asking for the dtypes of the relevant set of keys.

Pushed an update.